### ISSUE TITLE
Add naughty.replace_icon method

### ISSUE THREAD

**Issue by timroes** [awesomeWM/awesome#1731]
It would be very handy if there would be a method to also replace the icon of an existing notification and not just its text (as `naughty.replace_text` does.

*Current workaround:* Once can use `replaces_id` to replace the old notification if you need to change an icon in an notification. This has the drawback of a short flickering of the notification.

**Comment by blueyed** [comment]
Makes sense.
For this the `create iconbox` code would need to be factored out probably.

btw: it looks like using `replace_text` does not re-calculate the size/width according to the new text?!  (the `calculate the width` code)

**Comment by timroes** [comment]
The width issue is described by someone else in #1707.

Perhaps there should just be a general method to replace a notification and it should automatically determine what have changed, replace that parts of the notification (and maybe as a fallback, if there are things that can't easily be replaced, use the `replaces_id` mechanism).

Also I think `replaces_id` is often used in the same way, to have a place in code, that always do something like "show or replace" a notification of the same type. So meaning everyone does mainly the same steps:

1. Store a global variable, e.g. `last_id`
2. If that isn't `nil` set `replaces_id` to it
3. Store the returned notification's id to that `last_id` variable.

I build a mechanism around naughty, that just offers a `show_or_update` method, that just accepts a string key, so you could do something like the following:

```lua
notify('volume.control', {
  title = 'Volume',
  text = '[xxxx    ]',
  icon = 'audio-control-high'
})
```

You can use the very same call now, and it shows a new notification if none for the key 'volume.control' exists or updates that one. Also it does the part that I explained above, that it checks, whether it can use the existing `replaces_text` methods if only the text changes and `replaces_id` if something else changes.

Maybe an API with a - let's call it - grouping key would make sense in the naughty core? So that this common use-case of show or replace is covered by one single API, without the user needs to check themselves whether they need to call `replaces_text`, `replaces_icon`, or `replaces_id`. What are your thoughts on that?

**Comment by bew** [comment]
> This has the drawback of a short flickering of the notification.

Note that this is not longer true it has been fixed in #1785.

**Comment by Elv13** [comment]
As of 15 minutes ago, there is now an API to define custom notification widgets. It is now also possible to get the notification object and do `my_notif_object.icon = "new_icon"`.
